File: 2023/day3/main.py
from typing import Iterable
from dataclasses import dataclass


@dataclass
class NumberLocation:
    value: int
    row: int
    start: int
    end: int


@dataclass
class SymbolLocation:
    row: int
    col: int


def read_input(file: str) -> Iterable[str]:
    with open(file, mode="r") as f:
        out = f.readlines()
        out = [line.replace("\n", "") for line in out]
    return out


def extract_numbers(input: Iterable[str]) -> Iterable[NumberLocation]:
    for i, line in enumerate(input):
        start = None
        for j, c in enumerate(line):
            if start is None and c.isdigit():
                start = j
            if start is not None:
                if not c.isdigit():
                    end = j
                    yield NumberLocation(int(line[start:end]), i, start, end)
                    start = None
                elif j == len(line) - 1:
                    end = j + 1
                    yield NumberLocation(int(line[start:end]), i, start, end)


def extract_symbols(input: Iterable[str]) -> Iterable[SymbolLocation]:
    for i, line in enumerate(input):
        for j, value in enumerate(line):
            if value != "." and not value.isdigit():
                yield SymbolLocation(i, j)


def adjacent(symbol: SymbolLocation, part: NumberLocation) -> bool:
    return (
        part.row >= symbol.row - 1
        and part.row <= symbol.row + 1
        and part.start <= symbol.col + 1
        and part.end >= symbol.col
    )


def is_part(part_loc: NumberLocation, symbols: Iterable[SymbolLocation]) -> bool:
    return any(adjacent(symbol, part_loc) for symbol in symbols)


def gear_ratio(
    symbols: Iterable[SymbolLocation], parts: Iterable[NumberLocation]
) -> Iterable[int]:
    for symbol in symbols:
        adjacent_values = [p.value for p in parts if adjacent(symbol, p)]
        if len(adjacent_values) == 2:
            yield adjacent_values[0] * adjacent_values[1]


def main() -> None:
    input = read_input("input.txt")
    symbols = [s for s in extract_symbols(input)]
    part_numbers = [x for x in extract_numbers(input) if is_part(x, symbols)]
    part1 = sum(x.value for x in part_numbers)
    print(f"Part 1: {part1}")

    part2 = sum(gear_ratio(symbols, part_numbers))
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()

File: 2023/day3/src/main.rs
use std::fs::read_to_string;

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

#[derive(Debug)]
struct Number {
    value: u32,
    row: usize,
    start: usize,
    end: usize,
}

#[derive(Debug)]
struct Symbol {
    row: usize,
    col: usize,
}

fn extract_numbers(input: &[String]) -> Vec<Number> {
    let mut out: Vec<Number> = Vec::new();
    let n_col = input[0].len();
    let mut _start: Option<usize> = None;
    for (row, line) in input.iter().enumerate() {
        _start = None;
        for (j, c) in line.chars().enumerate() {
            if _start.is_none() && c.is_ascii_digit() {
                _start = Some(j);
            }
            if _start.is_some() && (!c.is_ascii_digit() || j + 1 == n_col) {
                let start = _start.unwrap();
                let end = if !c.is_ascii_digit() { j } else { j + 1 };
                let value: u32 = line[start..end].parse().unwrap();
                let number = Number {
                    value,
                    row,
                    start,
                    end,
                };
                out.push(number);
                _start = None;
            }
        }
    }
    out
}

// TODO: can probably do this with iterator traits
fn extract_symbols(input: &[String]) -> Vec<Symbol> {
    let mut out: Vec<Symbol> = Vec::new();
    for (row, line) in input.iter().enumerate() {
        for (col, value) in line.chars().enumerate() {
            if value != '.' && !value.is_ascii_digit() {
                out.push(Symbol { row, col });
            }
        }
    }
    out
}
fn adjacent(number: &Number, symbol: &Symbol) -> bool {
    (number.row >= symbol.row - 1)
        && (number.row <= symbol.row + 1)
        && (number.start <= symbol.col + 1)
        && (number.end >= symbol.col)
}

fn gear_ratio(symbol: &Symbol, numbers: &[Number]) -> Option<u32> {
    let adjacent_numbers: Vec<&Number> = numbers.iter().filter(|n| adjacent(n, symbol)).collect();
    if adjacent_numbers.len() == 2 {
        return Some(adjacent_numbers[0].value * adjacent_numbers[1].value);
    };
    None
}

fn main() {
    let input = read_input("input.txt");
    let numbers = extract_numbers(&input);
    let symbols = extract_symbols(&input);
    let part1: u32 = numbers
        .iter()
        .filter(|n| symbols.iter().any(|s| adjacent(n, s)))
        .map(|n| n.value)
        .sum();

    println!("Part 1: {}", part1);

    let part2: u32 = symbols.iter().flat_map(|s| gear_ratio(s, &numbers)).sum();

    println!("Part 2: {}", part2);
}
